service/view_building_coordinator.cc

```diff
@@ -0,0 +1,183 @@
+#include "seastar/coroutine/maybe_yield.hh"
+#include "service/raft/group0_state_machine.hh"
+#include "service/raft/raft_group0.hh"
+#include "service/raft/raft_group0_client.hh"
+#include "utils/assert.hh"
+#include "utils/log.hh"
+#include "seastar/core/abort_source.hh"
+#include "seastar/core/with_scheduling_group.hh"
+#include "service/migration_manager.hh"
+#include "replica/database.hh"
+#include "view_info.hh"
+
+#include "service/view_building_coordinator.hh"
+
+static logging::logger vbc_logger("view_building_coordinator");
+
+namespace service {
+
+namespace vbc {
+
+view_building_coordinator::view_building_coordinator(abort_source& as, replica::database& db, raft_group0& group0, db::system_keyspace& sys_ks, const topology_state_machine& topo_sm) 
+    : _db(db)
+    , _group0(group0)
+    , _sys_ks(sys_ks)
+    , _topo_sm(topo_sm)
+    , _as(as) 
+{}
+
+future<group0_guard> view_building_coordinator::start_operation() {
+    auto guard = co_await _group0.client().start_operation(_as);
+    co_return std::move(guard);
+}
+
+future<> view_building_coordinator::await_event() {
+    _as.check();
+    co_await _cond.when();
+    vbc_logger.debug("event awaited");
+}
+
+future<view_building_coordinator::vbc_state> view_building_coordinator::load_coordinator_state() {
+    auto tasks = co_await _sys_ks.get_view_building_coordinator_tasks();
+    co_return vbc_state {
+        .tasks = std::move(tasks)
+    };
+}
+
+table_id view_building_coordinator::get_base_id(const view_name& view_name) {
+    return _db.find_schema(view_name.first, view_name.second)->view_info()->base_id();
+}
+
+future<> view_building_coordinator::run() {
+    auto abort = _as.subscribe([this] noexcept {
+        _cond.broadcast();
+    });
+
+    while (!_as.abort_requested()) {
+        vbc_logger.debug("coordinator loop iteration");
+        try {
+            auto state_opt = co_await update_coordinator_state();
+            if (!state_opt) {
+                // If state_opt is nullopt, it means there was work to do and the state has changed.
+                continue;
+            }
+            // TODO: Do actual work, send RPCs to build a particular view's range
+            co_await await_event();
+        } catch (...) {
+            
+        }
+        co_await coroutine::maybe_yield();
+    }
+}
+
+future<std::optional<view_building_coordinator::vbc_state>> view_building_coordinator::update_coordinator_state() {
+    vbc_logger.debug("update_coordinator_state()");
+
+    auto guard = co_await start_operation();
+    std::vector<canonical_mutation> cmuts;
+
+    auto state = co_await load_coordinator_state();
+    auto views = _db.get_views() 
+            | std::views::transform([] (view_ptr& v) { return view_name{v->ks_name(), v->cf_name()}; })
+            | std::ranges::to<std::vector>();
+    auto built_views = co_await _sys_ks.load_built_tablet_views();
+
+    if (auto to_add = get_views_to_add(state, views, built_views); !to_add.empty()) {
```

Comment:
What if somebody removes a view and then re-creates it very quickly - with the same name but different schema? I'm not sure that `update_coordinator_state` is guaranteed to be executed after every schema operation on a view, so it might miss it.

Unless I'm missing something and the problem does not actually exist, I can see two ways out of this:

- Include the mutations generated by `get_views_to_add`/`get_views_to_remove` in the same group0 operation as the one that does the schema change.
- Use table UUIDs to identify the views in the relevant tables. New tables have new, completely unique UUIDs so we avoid the ABA problem altogether.

Reply:
changed to using table_id